nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py

```diff
@@ -176,3 +177,108 @@ def _group_metric_calculate(self, group_metrics: Union[Dict[str, Tensor], List[T
         for group_metric in group_metrics:
             group_sum_metric += group_metric
         return group_sum_metric
+
+
+class AttentionSparsityAllocator(SparsityAllocator):
+    """
+    A specific allocator for attention has a structure like self-attention in Bert.
+    In this sparsity allocator, :math:`Q_{proj}`, :math:`K_{proj}`, :math:`V_{proj}` of the same attention head will in one group.
+    Sparsity allocator will combine their metrics and share the mask among these three weights in one group.
+    Note that by default, assuming QKV are implemented with `torch.nn.Linear`.
+    """
+    def __init__(self, pruner: Pruner, dim: int, dummy_input: Any):
+        super().__init__(pruner, dim=dim)
+        self.dummy_input = dummy_input
+
+    def _get_dependency(self):
+        """
+        Populate self.attention_name_groups by running inference on the module graph.
+        Currently, the group inferred AttentionWeightDependency is limited to a set of four weights, with the first
+        three corresponding to :math:`Q_{proj}`, :math:`K_{proj}`, :math:`V_{proj}` (in any order) and the last one being :math:`{Output}_{proj}`.
+        """
+        try:
+            graph = self.pruner.generate_graph(dummy_input=self.dummy_input)
+            dependency_tracer = AttentionWeightDependency(traced_model=graph.trace)
```

Comment:
The logic of `AttentionWeightDependency` is not robust enough, maybe we can implement a function to find the exact same subgraph with `SelfAttention`?

Reply:
you are right, I will check if `SelfAttention` implementation is similar in other transformers, and find a better way to match the attention.

Reply:
What is the constraint of `AttentionWeightDependency`, only support specific module type (such as only torch.nn.Module, not support F function) or also not support other transformer implementation (such as only support Huggingface)? May be we should provide more notes to help user notice these constraints.

Reply:
For checking several Huggingface transformer implementations, `bert`, `roberta`, `gpt_neo`, `bart` `bait`...have similar `Q K V Projection Structure`.
And implementations from other repos or even in Huggingface may not have the similar structure, so I think this mode can only support huggingface implementations right now.
I will add more information to explain these constraints.